### PR TITLE
Fix `String::Buffer` and `IO::Memory` capacity to grow beyond 1GB

### LINKED ISSUE #13987: Arithmetic Overflow on large HTTP response
Apologies if this has already been reported before, I wasn't able to find it.

I was using the following code to download a large file (1.1 GB).

```cr
require "http/client"

response = HTTP::Client.get "https://some.url.that.returns.a.large.file/"
```

Unfortunately I can't include the real URL because it's not publicly accessible but it's fairly easy to replicate. I got an `Arithmetic Overflow` exception with the following backtrace:

```
Unhandled exception: Arithmetic overflow (OverflowError)
  from /usr/lib/crystal/int.cr:464:31 in 'next_power_of_two'
  from /usr/lib/crystal/math/math.cr:742:5 in 'pw2ceil'
  from /usr/lib/crystal/string/builder.cr:46:26 in 'write'
  from /usr/lib/crystal/io.cr:568:11 in 'gets_to_end'
  from /usr/lib/crystal/http/client/response.cr:81:17 in 'consume_body_io'
  from /usr/lib/crystal/http/client/response.cr:105:11 in 'from_io?'
  from /usr/lib/crystal/http/client.cr:610:5 in 'exec_internal_single'
  from /usr/lib/crystal/http/client.cr:592:18 in 'exec_internal'
  from /usr/lib/crystal/http/client.cr:585:7 in 'exec'
  from /usr/lib/crystal/http/client.cr:721:5 in 'exec'
  from /usr/lib/crystal/http/client.cr:753:7 in 'exec'
  from /usr/lib/crystal/http/client.cr:410:3 in 'get'
```

Looking through the referenced code it appears the issue is `HTTP::Client` tries to resize the internal buffer it's using to store the response to the next largest power of 2. 1.1 GB is slightly more than 2³⁰ bytes so the next one up is 2³¹ or 2,147,483,648. That's 1 beyond the max of the signed 32 bit integer being used to store the size, resulting in an overflow and crash.

Is this intended behavior? Is there any way to switch Crystal to use 64 bit integers for memory/container sizes or any plans to change the default? 2³⁰ bytes isn't that high of a limit for some applications.

This occurred on Crystal 1.10.1.

### PR DESCRIPTION
**Pull request by straight-shoota**
The byte buffer for `IO::Memory` and `String::Buffer` grows exponentially by the next power of two bigger then the currently requested size. So for example, if space is requested for 78 bytes, it allocates for a capacity of 128 bytes (`1 << 7`).

The maximum size of a buffer is `Int32::MAX`, which is just one below a power of tw (`(1 << 31) - 1`). So if space is requested for more than `1 << 30` bytes, the next power of two would be `1 << 31` which is out of range and raises `ArithmeticError`.
This means, the sizes between `(1 << 30) + 1` and `(2 << 31) - 1` would all be valid but cannot be reached by the dynamically growing buffer. If the actually required bytesize does not exeed `(2 << 31) - 1`, it should be possible to allocate that much.

This patch clamps the maximum cacpacity to `Int32::MAX`. That enables sizes in that range and there will only be an error if that is definitely not enough space.

Resolves #13987

### CODE REVIEW

**Comment by straight-shoota** [comment]
I have no idea what's going on in wasm32 CI. The error seems definitely triggered by the changes in this PR (more precisely d227e8b737553d02f26a7efc5f7ae02c4881f4f3). But I don't see any direct connection to the failing code, which is in some time function in the spec framework (which runs after the example is completed).
No other platform shows any memory issues and I've double checked the code, so I'm quite certain the changes are legit.
Maybe it triggers a bug in wasm?

**Comment by straight-shoota** [comment]
This needs further investigation into the wasm runtime.
I'm skipping the >1 GB spec for `String::Builder` on wasm32 for now.

**Comment by ysbaddaden** [comment]
About the WASM32 issue: Strings are always terminated with a null byte (`\0`), so shouldn't the maximum size be Int32::MAX - 1? In fact the String class is special and it actually allocates bytesize + HEADER_SIZE + 1 (for the null byte), so maybe the maximum capacity in `String::Builder` should be `Int32 - String::HEADER_SIZE - 1`?

**Comment by straight-shoota** [comment]
Yeah, that's correct. This extra capacity for the final null byte should be taken care of in all capactity calculations, though. `String::Builder#initialize(capacity)` already adds `String::HEADER_SIZE + 1` to the `capacity` parameter, so maximum capacity should correctly be `Int32::MAX` because all extra requirements are already counted in.
And the spec works correctly in compiled mode building a string with a length of `Int32::MAX - String::HEADER_SIZE - 1` bytes.

**Comment by ysbaddaden** [comment]
The _initial_ buffer allocation accounts for HEADER_SIZE + 1 but the reallocations don't, though you likely account for it in the bytesize when you check for IO::EOFError.

BTW: a String bytesize can be exactly Int32::MAX + 1 for the null byte:

```crystal
slice = Slice(UInt8).new(Int32::MAX)
slice[-1] = 97

str = String.new(slice.to_unsafe, Int32::MAX)
p str[-1] # => 'a' (97)
```

I tried to make String::Builder always reallocate to bytesize + header size + 1, but I get crashes.

**Comment by ysbaddaden** [comment]
I reproduce the wasm32 error on my local with your branch. I wrote a patch that always allocates bytesize + header size + 1 and the error is gone. That really sounds like a buffer overflow.

I'm trying to apply it on your branch. So far I'm having spec errors because I can allocate up to Int32::MAX bytesize (yes, it's accepted in `String.new` for example) when your branch only allowed up to Int32::MAX - 1 (to account for the null byte in the bytesize).

**Comment by ysbaddaden** [comment]
The spec passes until we try to build the actual string object, which doesn't allocate anything, but merely writes a few things on the buffer (the trailing null byte and the string header).

The spec still succeeds, but calling `Time.monotonic` in `src/spec/example.cr:51` will always crash wasmtime. I tried to skip the eventual writes to the buffer, or doing anything, but it still crashes on the same call to `Time.mononotic` :sob:

**Comment by straight-shoota** [comment]
@ysbaddaden So do you think we can merge this?

**Comment by ysbaddaden** [comment]
Yes, the problem isn't your code change, but allocating that much memory in wasm seems to trigger an issue with the wasm libs.

The only detail related to the patch is that we can allocate _one more byte_ (see above comment) but what's 1 byte compared to 1GB :joy:

File: src/io/memory.cr [hunk 1]
```diff
@@ -458,6 +450,28 @@ class IO::Memory < IO
...
+  end
+
+  private def calculate_new_capacity(new_bytesize : Int32)
+    # If the new bytesize is bigger than 1 << 30, the next power of two would
+    # be 1 << 31, which is out of range for Int32.
+    # So we limit the capacity to Int32::MAX in order to be able to use the
+    # range (1 << 30) < new_bytesize < Int32::MAX
+    return Int32::MAX if new_bytesize > 1 << 30
+
+    Math.pw2ceil(new_bytesize)
```
Comment (beta-ziliani) [inline src/io/memory.cr:472]: Unrelated, but I wonder if we shouldn't change the formula like we did for [array](https://github.com/crystal-lang/crystal/pull/11482)

Comment (beta-ziliani) [reply src/io/memory.cr:472]: And the converse is also true; that we should cap the max of array too

Comment (beta-ziliani) [reply src/io/memory.cr:472]: Maybe time for a `Growable` module...

**Review by beta-ziliani — APPROVED** [review]

File: src/io/memory.cr (hunk 1 above)
Comment (straight-shoota) [reply src/io/memory.cr:472]: => #14173

**Pull request merged**